### PR TITLE
PP-5322: App user authentication

### PR DESCRIPTION
**Pull request by sfount**
Basic GitHub OAuth application authentication. 

Note -- this build should not be pushed to DockerHub, it does not consider running locally and will break anyone running Toolbox from `pay local`

### CODE REVIEW

File: src/web/errorHandler.ts [hunk 1]
```diff
@@ -46,11 +46,18 @@ const handleDefault = function handleDefault(
...
 ): void {
+  logger.error('Unhandled error caught by middleware stack')
+  logger.error(error.stack)
+
   if (res.headersSent) {
     return next(error)
   }
+  if (!req.isAuthenticated()) {
+    // don't render application structure for non authenticated issues
+    return res.redirect('/auth/unauthorised?default_throw')
```
Comment (jankowiakmaria) [inline src/web/errorHandler.ts]: is the query parameter being used somehow or is it just for indication in url?

File: src/lib/auth/github.spec.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  })
+
+  it('invokes callback with `false` invalid permissions given invalid permissions', async () => {
+    const spy = chai.spy()
+    mockery.registerMock('./../permissions', { isPermittedUser: invalidPermissions })
+    // eslint-disable-next-line prefer-destructuring
+    handleGitHubOAuthSuccessResponse = require('./github').handleGitHubOAuthSuccessResponse
+
+    await handleGitHubOAuthSuccessResponse('some-access-token', 'some-refresh-token', profile, spy)
+    expect(spy).to.have.been.called.once.with(null, false)
```
[the remark this answers is not in the document]
Comment (sfount) [reply src/lib/auth/github.spec.js]: Will update the name of the spy! The `called` method asserts that at least these values have been passed to the method called, as we have no control over what assertion is thrown by the gubbins I think it's sufficient to uniquely identify this call with null + validated (false in this test case)

File: src/lib/permissions.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  const team = config.auth.AUTH_GITHUB_TEAM_ID
+  return validateUserTeamMembership(user, token, team)
+}
+
+const isAdminUser = function isAdminUser(user, token) {
+  const team = config.auth.AUTH_GITHUB_ADMIN_TEAM_ID
+  return validateUserTeamMembership(user, token, team)
+}
+
+module.exports = { isPermittedUser, isAdminUser }
```
[the remark this answers is not in the document]
Comment (sfount) [reply src/lib/permissions.js]: Good point! Have moved `github.js` -> `github/strategy.js` and `permissions.js` to `github/permissions.js` to be explicit and allow additional integrations later

File: src/web/errorHandler.ts (hunk 1 above)
Comment (sfount) [reply src/web/errorHandler.ts]: Just an indication, it should be logged before this point too, this should never really be hit so I'll leave it to differentiate.

**Comment by sfount** [comment]
@jankowiakmaria thanks for the comments! Changes made https://github.com/alphagov/pay-toolbox/pull/120/commits/5634ef002b718e056a72f0899596fcfc119493e9

File: src/lib/auth/github.spec.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (jankowiakmaria) [reply src/lib/auth/github.spec.js]: I see. I guess we don't care about the exception that much (in the assertion).

File: src/web/errorHandler.ts (hunk 1 above)
Comment (jankowiakmaria) [reply src/web/errorHandler.ts]: 👍

**Pull request merged**